File: ocaml/99/12.ml
(* Decode a run-length encoded list. (medium)

Given a run-length code list generated as specified in the previous problem, construct its uncompressed version.

Solution

# decode [Many (4,"a"); One "b"; Many (2,"c"); Many (2,"a"); One "d"; Many (4,"e")];;
- : string list =
["a"; "a"; "a"; "a"; "b"; "c"; "c"; "a"; "a"; "d"; "e"; "e"; "e"; "e"]
 *)

type 'a rle =
  | One of 'a
  | Many of int * 'a ;;

let rec decode = function
  | [] -> []
  | One x :: t -> x :: decode t
  | Many (n,x) :: t when n > 0 -> x :: decode (Many (n-1,x) :: t)
  | Many (_,x) :: t -> decode t ;;

File: prolog/99problems/12.pl
% P12 (**): Decode a run-length compressed list.

% decode(L1,L2) :- L2 is the uncompressed version of the run-length
%    encoded list L1.
%    (list,list) (+,?)

decode([],[]).
decode([X|S], [X|T]) :- atom(X), decode(S,T).
decode([[X,N]|T], [X|S]) :- N > 0, Np is N-1, decode([[X,Np]|T], S).
decode([[X,0]|T], S) :- decode(T,S).

